sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Security/Ucfg/UniversalControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,331 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Protobuf.Ucfg;
+using SonarAnalyzer.SymbolicExecution.ControlFlowGraph;
+using UcfgLocation = SonarAnalyzer.Protobuf.Ucfg.Location;
+
+namespace SonarAnalyzer.Security.Ucfg
+{
+    public class UniversalControlFlowGraphBuilder
+    {
+        public static readonly string Const = "\"\""; // Same as Java implementation
+        private readonly BlockIdMap blockId = new BlockIdMap();
+        private readonly SemanticModel semanticModel;
+        private readonly InstructionBuilder instructionBuilder;
+        private readonly IControlFlowGraph cfg;
+
+        public UniversalControlFlowGraphBuilder(SemanticModel semanticModel, IControlFlowGraph cfg)
+        {
+            this.semanticModel = semanticModel;
+            instructionBuilder = new InstructionBuilder(semanticModel);
+            this.cfg = cfg;
+        }
+
+        public IEnumerable<BasicBlock> Build() =>
+            cfg.Blocks.Where(IsSupportedBlock).Select(CreateBasicBlock);
+
+        public BasicBlock CreateBasicBlock(Block block)
+        {
+            var basicBlock = new BasicBlock
+            {
+                Id = blockId.Get(block),
+            };
+
+            foreach (var instruction in block.Instructions)
+            {
+                instructionBuilder.BuildInstruction(instruction, basicBlock);
+            }
+
+            if (block is JumpBlock jump)
+            {
+                instructionBuilder.BuildInstruction(jump.JumpNode, basicBlock);
+            }
+            else
+            {
+                if (block.SuccessorBlocks.All(cfg.ExitBlock.Equals))
+                {
+                    basicBlock.Ret = new Return { ReturnedExpression = new Expression { Const = new Constant { Value = Const } } };
+                }
+                else
+                {
+                    basicBlock.Jump = new Jump();
+                    basicBlock.Jump.Destinations.AddRange(block.SuccessorBlocks.Select(blockId.Get));
+                }
+            }
+
+            return basicBlock;
+        }
+
+        private static bool IsSupportedBlock(Block block) =>
+            !(block is ExitBlock);
+
+        public static UcfgLocation GetLocation(SyntaxNode syntaxNode)
+        {
+            var location = syntaxNode.GetLocation();
+            var lineSpan = location.GetLineSpan();
+            return new UcfgLocation
+            {
+                FileId = location.SourceTree.FilePath,
+                StartLine = lineSpan.StartLinePosition.Line,
```

Comment:
Is the UCFG positioning zero-based?

Reply:
No idea, this is worth checking though.

Reply:
UCFG relies on lines starting at 1 and column starting at 1 as those are locations in the file and are as such starting with 1. (line 0 makes little sense). 
This question could (should) have been asked to me or @marcelo-sousa-sonarsource so we can improve our definition and documentation of the UCFG to help future developers to map a language to UCFG.

@valhristov can you please create a ticket to test and fix this accordingly and link it into this conversation ? 
So we can first merge the generation and then fix this ticket.
Thanks.

Reply:
https://jira.sonarsource.com/browse/SONARSEC-48